File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.py
class SmallestInfiniteSet:

    origin = set()
    smallest_origin = 1
    

    def __init__(self):
        for i in range(1, 1001):
            self.origin.add(i)
        
    
    def popSmallest(self) -> int:
        temp = self.smallest_origin
        
        # find next smallest in origin
        for i in range(temp + 1, 1001):
            if i in self.origin:
                self.smallest_origin = i
                break
        
        self.origin.remove(temp)
        
        return temp

    def addBack(self, num: int) -> None:
        self.origin.add(num)
        
        if num < self.smallest_origin:
            self.smallest_origin = num
        

File: 2336-smallest-number-in-infinite-set/2336-smallest-number-in-infinite-set.ts
class SmallestInfiniteSet {
  public smallest: number;
  public popped: Set<number>;

  constructor() {
    this.smallest = 1;
    this.popped = new Set();
  }

  popSmallest(): number {
    let temp = this.smallest;
    // console.log("curr smallest", temp);
    // console.log(this.popped);

    this.smallest += 1;
    while (this.popped.has(this.smallest)) {
      this.smallest += 1;
    }

    this.popped.add(temp);
    return temp;
  }

  addBack(num: number): void {
    if (!this.popped.has(num)) {
      return;
    }

    this.popped.delete(num);

    // if it contains, check if incomming num is actually smallest
    // console.log("num", num, "smallest", this.smallest);
    if (num < this.smallest) {
      // console.log("change");
      this.smallest = num;
    }
  }
}
